mlir/lib/QEC/Transforms/CommuteCliffordTPPR.cpp

```diff
@@ -0,0 +1,404 @@
+    auto lhsPauliStr = lhsPauli.str();
+    for (unsigned i = 0; i < rhsPauliStr.size(); i++) {
+        if (rhsPauliStr[i] == 'I' || rhsPauliStr[i] == '_') {
+            rhsPauliStr[i] = lhsPauliStr[i];
+        }
+    }
+    rhsPauli.pauliString = stim::PauliString<MAX_BITWORD>::from_str(rhsPauliStr.c_str());
+}
+
+SmallVector<Value> fullfillOperands(PauliStringWrapper lhsPauliWrapper,
+                                    PauliStringWrapper rhsPauliWrapper)
+{
+    auto lhs = lhsPauliWrapper.op;
+    auto rhs = rhsPauliWrapper.op;
+    auto rhsPauliSize = rhsPauliWrapper.correspondingQubits.size();
+
+    // Fullfill Operands of RHS
+    // lhsPauli.correspondingQubits consists of:
+    //    - OutQubit: initilized qubit
+    //    - InQubit: new qubits from RHS op
+    SmallVector<Value> newRHSOperands(rhsPauliSize, nullptr);
+    for (unsigned i = 0; i < rhsPauliSize; i++) {
+        for (unsigned j = 0; j < rhs.getInQubits().size(); j++) {
+            if (rhs.getInQubits()[j] == rhsPauliWrapper.correspondingQubits[i]) {
+                newRHSOperands[i] = rhs.getInQubits()[j];
+            }
+        }
+        for (unsigned j = 0; j < lhs.getOutQubits().size(); j++) {
+            if (lhs.getOutQubits()[j] == rhsPauliWrapper.correspondingQubits[i]) {
+                newRHSOperands[i] = lhs.getInQubits()[j];
+            }
+        }
+    }
+    return newRHSOperands;
+}
+
+SmallVector<StringRef> removeIdentityPauli(PPRotationOp rhs, SmallVector<Value> &newRHSOperands)
+{
+    auto pauliProduct = rhs.getPauliProduct();
+    SmallVector<StringRef> pauliProductArrayRef;
+
+    for (auto [i, pauli] : llvm::enumerate(pauliProduct)) {
+        auto pauliStr = mlir::cast<mlir::StringAttr>(pauli).getValue();
+        if (pauliStr == "I" || pauliStr == "_") {
+            newRHSOperands.erase(newRHSOperands.begin() + i);
+            continue;
+        }
+        pauliProductArrayRef.push_back(pauliStr);
+    }
+
+    return pauliProductArrayRef;
+}
+
+void moveCliffordPastNonClifford(PauliStringWrapper lhsPauli, PauliStringWrapper rhsPauli,
+                                 PauliStringWrapper *result, PatternRewriter &rewriter)
+{
+    assert(lhsPauli.op != nullptr && "LHS Operation is not found");
+    assert(rhsPauli.op != nullptr && "RHS Operation is not found");
+
+    auto lhs = lhsPauli.op;
+    auto rhs = rhsPauli.op;
+
+    assert(!isNonClifford(lhs) && "LHS Operation is not Clifford");
+    assert(isNonClifford(rhs) && "RHS Operation is not non-Clifford");
+    assert(lhs.getPauliProduct().size() == lhs.getOutQubits().size() &&
+           "LHS Pauli product size mismatch before commutation.");
+    assert(rhs.getPauliProduct().size() == rhs.getInQubits().size() &&
+           "RHS Pauli product size mismatch before commutation.");
+
+    // Update Pauli words of RHS
+    if (result != nullptr) {
+        updatePauliProduct(rhs, *result, rewriter);
+    }
+    else {
+        replaceIdentityPauli(rhsPauli, lhsPauli);
```

Comment:
I don't think I understand this, why do we replace identities in the rhs PPR with elements from the lhs PPR?